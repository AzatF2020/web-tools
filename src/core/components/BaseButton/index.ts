export { default as BaseButton } from './ui/BaseButton.tsx';
