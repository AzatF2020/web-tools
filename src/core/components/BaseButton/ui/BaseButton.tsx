import type { FC } from 'react';
import type { IButtonProps } from '@/core/components/BaseButton/types/types';
import styles from 'core/components/BaseButton/styles/BaseButton.module.scss';
import clsx from 'clsx';

const BaseButton: FC<IButtonProps> = ({
  size = 'md',
  variant = 'primary',
  type = 'button',
  loading = false,
  label,
  children,
  ...props
}) => {
  return (
    <button
      type={type}
      className={clsx(
        styles.button,
        styles[`button--${size}`],
        styles[`button--${variant}`],
        loading && styles['button--loading'],
      )}
      {...props}>
      <span className={styles.button__label}>{label || children}</span>
    </button>
  );
};

export default BaseButton;
