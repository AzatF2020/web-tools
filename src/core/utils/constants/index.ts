const sizes = ['sm', 'md', 'lg'] as const;
const variants = ['primary', 'secondary', 'link', 'outlined'];

export { sizes, variants };
