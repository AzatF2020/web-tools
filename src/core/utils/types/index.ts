type TSize = 'sm' | 'md' | 'lg';

type TVariant = 'primary' | 'secondary' | 'link' | 'outlined' | 'loading';

export type { TSize, TVariant };
