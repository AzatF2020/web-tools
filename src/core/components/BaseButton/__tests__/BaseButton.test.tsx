import BaseButton from '../ui/BaseButton';
import type {
  IBaseButtonProps,
  TButtonSize,
  TButtonVariant,
} from '@/core/components/BaseButton/types/types';
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { sizes, variants } from '@/core/utils/constants';

const renderButton = (props: IBaseButtonProps = { label: 'button' }) => {
  return render(<BaseButton {...props} />);
};

describe('Button test component', () => {
  it('should render component with props', () => {
    renderButton({
      size: 'lg',
      variant: 'primary',
      type: 'submit',
      label: 'Click on me',
      disabled: false,
    });
    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  describe('test prop size', () => {
    test.each(sizes)('render with prop size %s', (size) => {
      renderButton({ size: size as TButtonSize });
      expect(screen.getByRole('button')).toHaveClass(`button--${size}`);
    });
  });

  describe('test prop variant', () => {
    test.each(variants)('render with prop variant %s', (variant) => {
      renderButton({ variant: variant as TButtonVariant });
      expect(screen.getByRole('button')).toHaveClass(`button--${variant}`);
    });
  });

  it('should render component', () => {
    renderButton();
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('should render attribute disabled', () => {
    renderButton({ disabled: true });
    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('should render with right prop label', () => {
    renderButton({ label: 'Click on me' });
    expect(screen.getByText('Click on me')).toBeInTheDocument();
  });

  it('should render with empty prop label', () => {
    renderButton({ label: '' });
    expect(screen.getByRole('button').textContent).toHaveLength(0);
  });

  it('calls onClick prop when clicked', () => {
    const onClick = vi.fn();
    renderButton({ onClick });
    screen.getByRole('button').click();
    expect(onClick).toHaveBeenCalled();
  });

  it('calls onClick prop when clicked', () => {
    const onClick = vi.fn();
    renderButton({ onClick });
    screen.getByRole('button').click();
    expect(onClick).toHaveBeenCalledOnce();
  });
});
