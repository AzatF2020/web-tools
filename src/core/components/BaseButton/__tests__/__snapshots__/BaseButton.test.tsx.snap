// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button test component > should render component with props 1`] = `
<button
  class="button button--lg button--primary"
  type="submit"
>
  <span
    class="button__label"
  >
    Click on me
  </span>
</button>
`;
